src/nbl/asset/utils/CCompilerSet.cpp

```diff
@@ -0,0 +1,66 @@
+// Copyright (C) 2018-2022 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#include "nbl/asset/utils/CCompilerSet.h"
+
+using namespace nbl;
+using namespace nbl::asset;
+
+core::smart_refctd_ptr<ICPUShader> CCompilerSet::compileToSPIRV(const ICPUShader* shader, const IShaderCompiler::SCompilerOptions& options) const
+{
+	core::smart_refctd_ptr<ICPUShader> outSpirvShader = nullptr;
+	if (shader)
+	{
+		switch (shader->getContentType())
+		{
+		case IShader::E_CONTENT_TYPE::ECT_HLSL:
+		{
+			const char* code = reinterpret_cast<const char*>(shader->getContent()->getPointer());
+			outSpirvShader = m_HLSLCompiler->compileToSPIRV(code, options);
+		}
+		break;
+		case IShader::E_CONTENT_TYPE::ECT_GLSL:
+		{
+			const char* code = reinterpret_cast<const char*>(shader->getContent()->getPointer());
+			outSpirvShader = m_GLSLCompiler->compileToSPIRV(code, options);
+		}
+		break;
+		case IShader::E_CONTENT_TYPE::ECT_SPIRV:
+		{
+			outSpirvShader = core::smart_refctd_ptr<ICPUShader>(const_cast<ICPUShader*>(shader));
+		}
+		break;
+		}
+	}
+	return outSpirvShader;
+}
+
+core::smart_refctd_ptr<ICPUShader> CCompilerSet::preprocessShader(const ICPUShader* shader, const IShaderCompiler::SPreprocessorOptions& preprocessOptions) const
+{
+	if (shader)
+	{
+		switch (shader->getContentType())
+		{
+		case IShader::E_CONTENT_TYPE::ECT_HLSL:
+		{
+			const char* code = reinterpret_cast<const char*>(shader->getContent()->getPointer());
+			auto stage = shader->getStage();
+			auto resolvedCode = m_HLSLCompiler->preprocessShader(code, stage, preprocessOptions);
+			return core::make_smart_refctd_ptr<ICPUShader>(resolvedCode.c_str(), stage, IShader::E_CONTENT_TYPE::ECT_HLSL, std::string(shader->getFilepathHint()));
+		}
+		break;
+		case IShader::E_CONTENT_TYPE::ECT_GLSL:
+		{
+			const char* code = reinterpret_cast<const char*>(shader->getContent()->getPointer());
+			auto stage = shader->getStage();
+			auto resolvedCode = m_GLSLCompiler->preprocessShader(code, stage, preprocessOptions);
+			return core::make_smart_refctd_ptr<ICPUShader>(resolvedCode.c_str(), stage, IShader::E_CONTENT_TYPE::ECT_GLSL, std::string(shader->getFilepathHint()));
+		}
+		break;
+		default:
+			return nullptr;
```

Comment:
@Erfan-Ahmadi @deprilula28 don't you think we should return the original shader if its SPIR-V ?

Reply:
Oh definitely, I wrote it this way for compilation function `CCompilerSet::compileToSPIRV`, we missed it for the preprocess function.
@deprilula28 please fix this issue and open a PR for these HLSL fixes.

------------------
2. `CCompilerSet::getShaderCompiler` will return nullptr for SPIRV language or anything other than HLSL/GLSL. so Add nullptr check guards when you're using the compiler directly (to get includeFinder) using this funciton (in CVulkanLogicalDevice) 

NOTE: we could make a "mock" spirv compiler but since the IShaderCompiler compile functions take a `const char * code` right now and spirv is not null terminated it'll be a little time consuming to make it work. just make sure we handle the case where it returns nullptr.